### PR TITLE
Implement infrastructure fro a `use` keyword in `*.wit`

### PR DESCRIPTION
**Pull request by alexcrichton**
This PR is not ready for review yet but I wanted to open it up to help prepare for the scale of change happening here. This is becoming a very large change, which is somewhat expected, and it's also not done yet! The current state of this PR is:

* An AST intermediate format is created to represent `use`, basically moving a number of arenas into a `Document`. More-or-less the representation changes here are fairly simple and while they have a lot of knock-on effects it's relatively mechanical.
* Support is added to `wit-component` to encode `use` statements in a component and additionally decode it. This includes translation of a `use`-using document into binary and back again as well as printing to the text format.
* Support for parsing the text format is _not_ included yet. @fibonacci1729 is working on that in parallel which will encompass pieces such as loading from the filesystem, resolution of a document, etc.
* A simple fuzzer has been added to generate documents (in-memory, not text) and then work with them with `wit-component`. This is what I've primarily been using to develop this PR and there are still a number of outstanding issues.

There are quite a few changes here in this PR but my personal conclusion at least is that `*.wit` is still early-enough in its development that this scale of PR won't be helped much by breaking it up, mostly requiring a lot of extra work to try to keep everything working in the interim.

My rough plan is to continue fixing fuzz-bugs and working with @fibonacci1729 on the text format this week. We'll merge the text format into here and then ensure there are plenty of tests for `use` both for `wit-parser` and for `wit-component`. At the moment there are no tests other than the fuzzer since the text parser isn't implemented yet.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Ok I talked with Luke about  the remaining major feature here, which is `wit-component` support for an `export` interface which `use`s a type from another `export` interface. The short version is that the component model binary format doesn't give a way right now to encode this information easily in the types-only mode (and it's a bit wonky in "real"  mode too). The changes to the binary format coming with the introduction of `resource` types will fix this, however. My current plan is to leave this largely unimplemented where it'll just panic for now within the `wit-component` phases. This should be ok for now as it's not expected to come up with idiomatic wit for the time being and should be implemened by the time it does come up.

Otherwise in terms of staging for this PR, I think it may actually be reasonable to go ahead and land this. This doesn't break any existing use cases of `wit-component` and gets the AST in a state which host/guest generators can start updating to. @fibonacci1729's work to get the text parsing will make this all actually usable but should slot in nicely to the `wit-parser` crate and into the AST structure created here.

@peterhuene or @pchickey I realize this is a very large PR, would one of y'all be willing to review it? If you'd like I'm happy to talk over video about this and walk through the higher-level portions as well.

**Comment by peterhuene** [comment]
I'll try to tackle this PR today and will ping you if a chat is necessary.

File: crates/wit-component/tests/interfaces/reference-out-of-order/world.wit [hunk 1]
```diff
@@ -3,14 +3,14 @@ interface foo {
...
 
-  record r-no-string {
-    s: u32,
-  }
-
   variant v {
     s(string),
   }
 
+  record r-no-string {
```
Comment (peterhuene) [inline crates/wit-component/tests/interfaces/reference-out-of-order/world.wit:10]: Note to self / other reviewers: the order here changed because encoding now encodes types as encountered in function signatures rather than first passing over the types in the interface to encode. Thus it's `r`, `v`, `r-no-string`, `v-no-string` now.

**Review by peterhuene — COMMENTED** [review]
I'm 80 / 98 files reviewed and stopped at the wit-parser changes, which I'll resume tomorrow to review.

File: crates/wit-component/fuzz/fuzz_targets/roundtrip-wit.rs [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+    let path = Path::new(path);
+    std::fs::write(path, contents).unwrap();
+    if path.extension().and_then(|s| s.to_str()) == Some("wasm") {
+        let path = path.with_extension("wat");
+        log::debug!("writing file {}", path.display());
+        std::fs::write(path, wasmprinter::print_bytes(&contents).unwrap()).unwrap();
+    }
+}
+
+mod generate {
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wit-component/fuzz/fuzz_targets/roundtrip-wit.rs:69]: While theoretically any component should be able to get a `*.wit` document extracted from it I think we still have aways to go for that on the specification. For example named types aren't implemented in such a way to allow that to infallibly happen. Additionally the compnent generation hasn't been updated in awhile so it may not produce the most interesting components to extract a document from.

You're right though in that I would rather generate the text rather than an AST directly here. For now this is just an intial state of a fuzzer though since I didn't have the text parsing on-hand. I hope to improve this and perhaps merge it into the top-level fuzz targets eventually if it's useful enough. For now though it's only really useful for the development of this PR so far (although will be helpful in future refactorings)

File: crates/wit-component/tests/interfaces/empty/types_only.wat [hunk 3]
```diff
@@ -1,10 +1 @@
...
-    (instance)
-  )
-  (type (;1;)
-    (instance)
-  )
-  (import "empty" (instance (;0;) (type 0)))
-  (export "empty" (type 1))
-)
\ No newline at end of file
+(component)
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wit-component/tests/interfaces/empty/types_only.wat:1]: I agree with you. This is an artifact of the fuzzer I wrote where the types-only document and the component-based document are required to be equal, but instead I think it would be better to go for a lossless representation of a document as a binary component. In https://github.com/WebAssembly/component-model/pull/141 I'm proposing a much different binary format than is currently emitted here but should be suitable for a lossless representation of a document.

Additionally my plan currently is to remove `--types-only` and instead have:

* `wasm-tools component wit component.wasm` - prints the WIT interface of an existing component
* `wasm-tools component wit foo.wit` - prints the "single string" version of a WIT document after resolving all local filesystem paths and such
* `wasm-tools component wit foo.wit --binary` - prints the binary format as specified in https://github.com/WebAssembly/component-model/pull/141 (with a standard `-t` flag for using the wasm text format)
* `wasm-tools component wit wit.wasm` - prints the WIT document for a binary-encoding wit file

Basically moving the "types only" mode to `wasm-tools component wit --binary <input-doc>` instead of part of `wasm-tools component new`. Bikeshedding is of course welcome on all this!

File: crates/wit-component/tests/interfaces/records/world.wit [hunk 4]
```diff
@@ -55,7 +55,7 @@ interface records {
 
 }
 
-interface records-exports {
+interface records0 {
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wit-component/tests/interfaces/records/world.wit:58]: I agree this isn't great and it's anothing thing the [lossles binary format above](https://github.com/bytecodealliance/wasm-tools/pull/858#discussion_r1043771153) I think should fix.

File: crates/wit-component/tests/interfaces/empty/types_only.wat (hunk 3 above)
[the remark this answers is not in the document]
Comment (peterhuene) [reply crates/wit-component/tests/interfaces/empty/types_only.wat:1]: Oh, I completely missed that there was an alternative encoding scheme proposed in that PR. Let me review the upstream PR to get up-to-speed, but :+1: on not being able to round-trip certain constructs until the encoding is updated.

[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wit-component/tests/interfaces/empty/types_only.wat:1]: Ah yeah this space has been moving a bit quickly and basically changed the moment I posted my own PR here. I plan to talk more about what's going on at the component-model meeting tomorrow.

File: crates/wit-parser/tests/ui/parse-fail/duplicate-value.wit.result [hunk 5]
```diff
@@ -1,5 +1,5 @@
-`a` is defined more than once
-     --> tests/ui/parse-fail/duplicate-value.wit:5:3
+expected '(', found keyword `s32`
```
Comment (peterhuene) [inline crates/wit-parser/tests/ui/parse-fail/duplicate-value.wit.result:1]: I think this test should now be deleted given we can't declare values this way in interfaces.

**Review by peterhuene — APPROVED** [review]
Looks good, just a few minor things after reviewing `wit-parser`.

Thanks for cleaning up a bit of the parser implementation as well (the span tracking in `Document` just for world/interface conversion I never liked despite implementing it).

**Comment by alexcrichton** [comment]
Thinking about this I am leaning towards merging this as-is and continuing to iterate from here. Thanks again @peterhuene for going through all of this!

I've been talking with Luke and Brian about the design space here and while we may not end up exactly on what's written here I think the refactorings are nevertheless good to have and beneficial. Namely the change in `wasm-tools component new` to encode types as-we-go rather than all up-front I think will be crucial for any form of `use`d types across interfaces, and that's a large chunk of this refactoring.

The next steps here are:

* Finish design work on https://github.com/WebAssembly/component-model/pull/141, namely what `use` can do, how to write it down, how it relates to packages, and what the binary format looks like
* Finish implementing the text parser and resolution in this repository to start adding tests (@fibonacci1729 is working on this)
* Implement the binary format specified upstream and refactor how this works on the CLI [as I mentioned above](https://github.com/bytecodealliance/wasm-tools/pull/858#discussion_r1043771153)

In the meantime I believe @fibonacci1729 is working on parsing/resolution of a single file with `use`-between-interfaces, so not multi-file yet, and that should open up being able to write interesting tests here.

**Comment by fibonacci1729** [comment]
@alexcrichton That makes sense to me!
